plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/wizard/application/TemplateListPage.java

```diff
@@ -69,11 +79,101 @@ public TemplateListPage(IWizard wizard, ITemplateListPageModel model) {
 	@Override
 	protected void doCreateControls(Composite parent, DataBindingContext dbc) {
 		GridLayoutFactory.fillDefaults()
-			.numColumns(1)
+			.numColumns(2)
 			.margins(10, 6)
 			.spacing(2, 2)
 			.applyTo(parent);
 		
+		// upload template
+		SelectObservableValue uploadTemplate = new SelectObservableValue();
+		ValueBindingBuilder
+			.bind(uploadTemplate)
+			.to(BeanProperties.value(
+					ITemplateListPageModel.PROPERTY_USE_UPLOAD_TEMPLATE).observe(model))
+			.in(dbc);
+		
+		createUploadControls(parent, uploadTemplate, dbc);
+		
+		createServerTemplateControls(parent, uploadTemplate, dbc);
+		model.setUseUploadTemplate(false);
+	}
+
+	private void createUploadControls(Composite parent, SelectObservableValue uploadTemplate, DataBindingContext dbc) {
+		// upload  app radio
+		Button btnUploadTemplate = new Button(parent, SWT.RADIO);
+		btnUploadTemplate.setText("Use a template from my local file system:");
+		GridDataFactory.fillDefaults()
+				.align(SWT.FILL, SWT.CENTER)
+				.span(2, 1)
+				.grab(true, false)
+				.applyTo(btnUploadTemplate);
+
+		uploadTemplate.addOption(Boolean.TRUE, WidgetProperties.selection().observe(btnUploadTemplate));
+
+		// uploaded file name
+		this.txtUploadedFileName = new Text(parent, SWT.BORDER);
+		txtUploadedFileName.setEnabled(false);
+		GridDataFactory.fillDefaults()
+				.align(SWT.FILL, SWT.CENTER)
+				.grab(true, false)
+				.applyTo(txtUploadedFileName);
+		IObservableValue uploadedFilenameTextObservable = WidgetProperties.text(SWT.Modify).observe(txtUploadedFileName);
+		ValueBindingBuilder
+				.bind(uploadedFilenameTextObservable)
+				.to(BeanProperties.value(
+						ITemplateListPageModel.PROPERTY_TEMPLATE_FILENAME).observe(model))
+				.in(dbc);
+
+		// browse button
+		Button btnBrowseFiles = new Button(parent, SWT.NONE);
+		btnBrowseFiles.setText("Browse...");
+		GridDataFactory.fillDefaults()
+				.align(SWT.LEFT, SWT.CENTER).hint(100, SWT.DEFAULT).grab(false, false)
+				.applyTo(btnBrowseFiles);
+		ValueBindingBuilder
+				.bind(WidgetProperties.enabled().observe(btnBrowseFiles))
+				.notUpdatingParticipant()
+				.to(uploadTemplate)
+				.in(dbc);
+		btnBrowseFiles.addSelectionListener(new SelectionAdapter() {
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				FileDialog dialog = new FileDialog(getShell(), SWT.OPEN);
+				if(StringUtils.isNotBlank(model.getTemplateFileName())) {
+					File file = new File(model.getTemplateFileName());
+					dialog.setFilterPath(file.getParentFile().getAbsolutePath());
+				}
+				String file = "";
+				while(file != null) {
+					file = dialog.open();
+					if(file != null) {
```

Comment:
@adietish maybe this should be a do/while since we want at least 1 pass through.  dialog.open only returns null when a user does not choose a file.  This logic was added to handle the case where a user picks a malformed template file, we notify, and allow them to choose another.